src/data/polynomial/iterated_deriv.lean

```diff
@@ -0,0 +1,169 @@
+/-
+Copyright (c) 2020 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import data.polynomial.derivative
+import logic.function.iterate
+import tactic.ring tactic.linarith
+
+/-!
+# Theory of iterated derivative
+We define and prove some lemmas about iterated (formal) derivative for polynomials over a semiring.
+-/
+
+noncomputable theory
+
+open finset nat polynomial
+open_locale big_operators
+
+namespace polynomial
+universes u
+variable {R : Type u}
+
+section semiring
+
+variables [semiring R] (f p q: polynomial R) (n k : ℕ)
+
+/--Defining the n-th derivative of polynomial f-/
+def iterated_deriv : polynomial R := derivative ^[n] f
+
+@[simp] lemma iterated_deriv_zero_right : iterated_deriv f 0 = f := rfl
+
+lemma iterated_deriv_succ : (iterated_deriv f (n+1)) = (iterated_deriv f n).derivative :=
+by rw [iterated_deriv, iterated_deriv, function.iterate_succ']
+
+@[simp] lemma iterated_deriv_zero : iterated_deriv (0 : polynomial R) n = 0 :=
+begin
+  induction n with n hn;
+  simp only [iterated_deriv, id.def, function.iterate_zero],
+  rwa ←iterated_deriv
+end
+
+@[simp] lemma iterated_deriv_add : (iterated_deriv (p+q) n) = (iterated_deriv p n) + (iterated_deriv q n) :=
+begin
+  induction n with n ih,
+  { simp only [iterated_deriv_zero_right], },
+  { rw [iterated_deriv_succ, ih, polynomial.derivative_add, iterated_deriv_succ, iterated_deriv_succ] }
+end
+
+end semiring
+
+section comm_ring
+variable [comm_ring R]
+variables (f p q: polynomial R) (n k : ℕ)
+
+
+lemma iterated_deriv_coeff :
+  ∀ m : ℕ, (iterated_deriv f k).coeff m = (∏ i in finset.range k, (m+k-i)) * (f.coeff (m+k)) :=
+begin
+  induction k with k ih,
+  { simp only [add_zero, forall_const, one_mul, range_zero, eq_self_iff_true,
+      iterated_deriv_zero_right, prod_empty] },
+  { intro m, rw iterated_deriv_succ,
+    rw [polynomial.coeff_derivative, ih (m+1), prod_range_succ],
+    simp only [nat.cast_succ, succ_eq_add_one],
+    conv_rhs { rw [mul_assoc, mul_comm] },
+    have triv : (∏ (i : ℕ) in range k, (m + 1 + k - i : R)) = ∏ (x : ℕ) in range k, (m + (k + 1) - x),
+    { apply congr_arg, ext, ring },
+    rw triv,
+    replace triv : (m + 1 : R) = (m + (k+1) - k:R),
+    { rw add_sub_assoc, simp only [add_sub_cancel'], }, rw ←triv,
+      replace triv : f.coeff (m + 1 + k) = f.coeff (m + (k + 1)),
+    { apply congr_arg, ring},
+    rw triv }
+end
```

Comment:
Here's a slightly more general statement and its proof. I swapped the order of `f.coeff` and `∏` to match `coeff_derivative`.

```suggestion
  ∀ m : ℕ, (iterated_deriv f k).coeff m = f.coeff (m + k) * (∏ i in finset.range k, ↑(m + k - i)) :=
begin
  induction k with k ih,
  { simp },

  intro m,
  calc (f.iterated_deriv k.succ).coeff m
      = f.coeff (m + k.succ) * (∏ i in finset.range k, ↑(m + k.succ - i)) * (m + 1) :
    by rw [iterated_deriv_succ, coeff_derivative, ih m.succ, succ_add, add_succ]
  ... = f.coeff (m + k.succ) * (↑(m + 1) * (∏ (i : ℕ) in range k, ↑(m + k.succ - i))) :
    by { push_cast, ring }
  ... = f.coeff (m + k.succ) * (∏ (i : ℕ) in range k.succ, ↑(m + k.succ - i)) :
    by { rw [prod_range_succ, nat.add_sub_assoc (le_succ k), nat.succ_sub le_rfl, nat.sub_self] }
end
```